src/ebuckets.c

```diff
@@ -1775,6 +1776,78 @@ void ebValidate(ebuckets eb, EbucketsType *type) {
         ebValidateRax(ebGetRaxPtr(eb), type);
 }
 
+eItem ebDefragItem(ebuckets *eb, EbucketsType *type, eItem item, ebDefragFunction *fn) {
+    assert(!ebIsEmpty(*eb));
+    if (ebIsList(*eb)) {
+        ExpireMeta *prevem = NULL;
+        eItem curitem = ebGetListPtr(type, *eb);
+        while (curitem != NULL) {
+            if (curitem == item) {
+                if ((curitem = fn(curitem))) {
+                    if (prevem)
+                        prevem->next = curitem;
+                    else
+                        *eb = ebMarkAsList(curitem);
+                }
+                return curitem;
+            }
+
+            /* Move to the next item in the list. */
+            prevem = type->getExpireMeta(curitem);
+            curitem = prevem->next;
+        }
+    } else {
+        rax *rax = ebGetRaxPtr(*eb);
```

Comment:
It is not required to iterate the rax from start. It is very expensive and redundant.
Enough to make local, direct adjustments to the item and its segment.

All the rax part should be something like:


```
        /* Handling rax segment of item directly. 
         * Find prev of `item` by iterating over all items in segment in cycle
         */
        ExpireMeta *mIter = type->getExpireMeta(item);
        CommonSegHdr *currHdr;
        while (mIter->lastInSegment == 0)
            mIter = type->getExpireMeta(mIter->next);
        if (mIter->lastItemBucket)
            currHdr = (CommonSegHdr *) mIter->next;
        else
            currHdr = (CommonSegHdr *) ((NextSegHdr *) mIter->next)->prevSeg;

        /* If the item is the first in the segment, then update the segment header */
        if (currHdr->head == item) {
            if ((item = fn(item))) {
                currHdr->head = item;
            }
            return item;
        }
        
        /* Iterate over all items in the segment until the next is 'item' */
        ExpireMeta *mHead = type->getExpireMeta(currHdr->head);
        mIter = mHead;
        while (mIter->next != item)
            mIter = type->getExpireMeta(mIter->next);

        if ((item = fn(item))) {
            mIter->next = item;
        }
        return item;
```

Now it is bounded to iterating no more than 16 items in a cycle.

Reply:
great idea.

Reply:
Please add assert to the last `while` in case reached end of segment without finding the `item`.

Reply:
done with [`c952c1e` (#13229)](https://github.com/redis/redis/pull/13229/commits/c952c1e8f2eca64c8a45b3b3148b65027c1ef546)